### PR TITLE
TRUNK-5010 Able to merge patients with discontinued orders

### PR DESCRIPTION
**Pull request by sidtharthanan**
## Related Issue

see https://issues.openmrs.org/browse/TRUNK-5010

## Checklist:

- [x] My pull request only contains one single commit.
- [ ] My pull request is based on the latest master branch
  `git pull --rebase upstream master`.
- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.
- [ ] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by dkayiwa** [comment]
@sidtharthanan did you run mvn clean install before committing? The formatting does not look correct. :)

**Comment by sidtharthanan** [comment]
I ran both the commands `mvn clean install` and `mvn clean package`.
I don't see any formatting changes after the `mvn clean...` commands.
If you could tell me the exact file, I can modify them.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 1]
```diff
@@ -592,7 +587,22 @@ public void mergePatients(Patient preferred, Patient notPreferred) throws APIExc
...
+	private void requireNoActiveOrderOfSameType(Patient patient1, Patient patient2) {
+		List<Order> ordersByPatient1 = Context.getOrderService().getAllOrdersByPatient(patient1);
+		List<Order> ordersByPatient2 = Context.getOrderService().getAllOrdersByPatient(patient2);
+		ordersByPatient1.forEach((Order order1)->{
+			ordersByPatient2.forEach((Order order2)->{
+				if(order1.isActive() && order2.isActive() && order1.getOrderType().equals(order2.getOrderType())){
+					log.debug("Merge operation cancelled: Cannot merge patient" + patient1.getPatientId()
+						+ " to patient" + patient2.getPatientId());
+					log.debug("Patients having active order of " + order1.getOrderType());
+					throw new APIException("Patient.cannot.merge", (Object[]) null);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Does the exception message reflect the new business logic?

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 2]
```diff
@@ -143,7 +144,17 @@ private void voidOrders(Collection<Patient> patientsWithOrders) {
 			}
 		}
 	}
-	
+
+	private void voidOrders(Collection<Patient> patients, OrderType ot) {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Something like voidOrdersForType or anything else along those lines would make this method more meaningful.

**Comment by dkayiwa** [comment]
@sidtharthanan did you take a look at the formatter here? https://wiki.openmrs.org/display/docs/Coding+Conventions

https://talk.openmrs.org/t/openmrs-code-formatter-styles/794

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java (hunk 2 above)
Comment (sidtharthanan) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Okay. I will do that.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 1 above)
Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Yes, It means that patients are having active order of <certain type>. Hence, cannot merge.

**Comment by sidtharthanan** [comment]
I read this: https://wiki.openmrs.org/display/docs/Coding+Conventions.
I think the code follows all of them.

I read this: https://talk.openmrs.org/t/openmrs-code-formatter-styles/794.
I don't get what to do.

I tried installing this following code formatter: https://github.com/suniala/openmrs-code-style-idea.
This formatter changes a lot of lines that I have not worked at all.
Could you tell me what formatter to use for intellij or command line?

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 1 above)
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Did you get a chance to look at the actual text that this code represents in the messages.properties file? :)

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Do I have to update the `"Patient.cannot.merge"` property just in the messages.properties file or all the *.properties files?
Some of those files have commented lines with `english text`.

Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I would be in favour of a new message code that is descriptive enough to reflect the message. For message properties files, if you do not know the equivalent text in other languages, then you do not need to touch their message properties files.

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: `Patient.merge.bothHasActiveOrders=Cannot merge patients. Both {0} and {1} patients have active {2} orders.`

ex: `Cannot merge patients. Both 'COM238821' and 'BAH238591' patients have active 'Drug' orders.`

Something like this. Is it fine?

**Comment by sidtharthanan** [comment]
@dkayiwa I have updated the PR as requested. Could check it?

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 3]
```diff
@@ -593,6 +588,23 @@ public void mergePatients(Patient preferred, Patient notPreferred) throws APIExc
...
+		List<Order> ordersByPatient1 = Context.getOrderService().getAllOrdersByPatient(patient1);
+		List<Order> ordersByPatient2 = Context.getOrderService().getAllOrdersByPatient(patient2);
+		ordersByPatient1.forEach((Order order1) -> {
+			ordersByPatient2.forEach((Order order2) -> {
+				if (order1.isActive() && order2.isActive() && order1.getOrderType().equals(order2.getOrderType())) {
+					Object[] parameters = { patient1.getPatientId(), patient2.getPatientId(), order1.getOrderType() };
+					String message = Context.getMessageSourceService().getMessage(messageKey, parameters,
+							Context.getLocale());
+					log.debug(message);
+					throw new APIException(message);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Why not just use this? new APIException(String messageKey, Object[] parameters)

File: api/src/main/resources/messages.properties [hunk 4]
```diff
@@ -1667,6 +1667,7 @@ Patient.merge.notPreferred=Not Preferred
 Patient.merge.includeVoided=Include deleted
 Patient.merged=Patients merged successfully
 Patient.merge.patient.NoUnvoidedOrders=Not preferred patient(s) should not have any unvoided orders
+Patient.merge.bothHasActiveOrders=Cannot merge patients. Both {0} and {1} patients have active [{2}] orders.
```
Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: I would expect the error message to also reflect the second condition under which you are throwing the exception. The same applies to the message key.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 3 above)
Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Could you elaborate on `this`?

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: `String message = Context.getMessageSourceService().getMessage(messageKey, parameters,
							Context.getLocale());
					log.debug(message);
					throw new APIException(message);`
We're logging the message.

File: api/src/main/resources/messages.properties (hunk 4 above)
Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: Could you tell me what is the second condition?
The logic is that if `both the patients are having active orders of same type` the error is thrown.

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: Let me ask it this way, would the merge fail if two patients simply have active orders?

Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: If both the patients are having active orders of **same type** the error is thrown.

Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: Lets say patient A has `1 Active Drug Order` and patient B has `1 Active Lab Order`.
We will allow to merge those patients.

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: Does the messages.properties changes reflect that? :)

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: I think i now see type reflected in the third parameter. :)

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: How about? Patient.merge.sameTypeActiveOrders

Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: An possible error message would be:
Cannot merge patients. Both BAH109245 and BAH239281 patients have active [Drug Order] orders.
I hope that message is clear.
It would be good, If you can tell a better message, If you have any.

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: Yes that makes perfect sense! I have just noticed that the third parameter takes care of the order type. :)

Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: Lets go with this: Patient.merge.bothHasSameTypeActiveOrders.
What do you think?

Comment (dkayiwa) [inline api/src/main/resources/messages.properties]: How about cannotHaveSameTypeActiveOrders? :)

Comment (sidtharthanan) [inline api/src/main/resources/messages.properties]: Okay. I'll go with that. `cannotHaveSameTypeActiveOrders`

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 5]
```diff
@@ -593,6 +588,23 @@ public void mergePatients(Patient preferred, Patient notPreferred) throws APIExc
 		Context.getPersonService().savePersonMergeLog(personMergeLog);
 	}
 	
+	private void requireNoActiveOrderOfSameType(Patient patient1, Patient patient2) {
+		String messageKey = "Patient.merge.cannotHaveSameTypeActiveOrders";
+		List<Order> ordersByPatient1 = Context.getOrderService().getAllOrdersByPatient(patient1);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: There is a getActiveOrders() that you can call and simplify this code

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 6]
```diff
@@ -3304,15 +3315,43 @@ public void purgePatientIdentifierType_shouldThrowErrorWhenTryingToDeleteAPatien
...
 	
 	/**
-	 * @verifies fail if not preferred patient has unvoided orders
+	 * @verifies fail if multiple patients has active order of same type
 	 * @see PatientService#mergePatients(org.openmrs.Patient, org.openmrs.Patient)
+	 * @throws APIException
 	 */
 	@Test
-	public void mergePatients_shouldFailIfNotPreferredPatientHasUnvoidedOrders() throws Exception {
+	public void mergePatients_shouldFailIfMultiplePatientsHasActiveOrderOfSameType() throws Exception {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: This method name is a little confusing, can you rephrase it please and actually it would be nice for it to mention drug order in the method name, how about `mergePatients_shouldFailIfBothPatientsHaveActiveDrugOrders`

File: api/src/main/resources/messages.properties [hunk 7]
```diff
@@ -1667,6 +1667,7 @@ Patient.merge.notPreferred=Not Preferred
 Patient.merge.includeVoided=Include deleted
 Patient.merged=Patients merged successfully
 Patient.merge.patient.NoUnvoidedOrders=Not preferred patient(s) should not have any unvoided orders
+Patient.merge.cannotHaveSameTypeActiveOrders=Cannot merge patients. Both {0} and {1} patients have active [{2}] orders.
```
Comment (wluyima) [inline api/src/main/resources/messages.properties]: You should have specific messages based on the error and not this generic one, you should have an error message when multiple patients have active drug orders

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 5 above)
Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: getActiveOrders() method **excludes** discontinued orders.
But, getAllOrdersByPatient **includes** discontinued orders.

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java [hunk 8]
```diff
@@ -593,6 +588,23 @@ public void mergePatients(Patient preferred, Patient notPreferred) throws APIExc
...
+		List<Order> ordersByPatient2 = Context.getOrderService().getAllOrdersByPatient(patient2);
+		ordersByPatient1.forEach((Order order1) -> {
+			ordersByPatient2.forEach((Order order2) -> {
+				if (order1.isActive() && order2.isActive() && order1.getOrderType().equals(order2.getOrderType())) {
```
Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Do you mean to say that:
Active 'Drug Order' of 'Same chemical' (for clarity) should not exist in both patients while merging?

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 5 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I thought that's the whole point of this ticket and logic, i.e to check if the patient has any active orders, discontinued orders are never active

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I guess the question here is do we also want to copy the medical history of the not preferred?

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Note that it's one thing to check if any of the patients has any active orders and it's another to copy over the orders, when checking if any of the patients has any active orders, getActiveOrders() is what you need to call, when moving over the orders to the preferred  patient, then you need to call getAllOrders, makes sense? In this method I believe you are checking if they have active orders and you should be calling getActiveOrder

File: api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java (hunk 8 above)
Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: [Could you have a look it](https://talk.openmrs.org/t/patient-merge-functionality/9310/16) and whether it clarifies some of your questions or not?
Specially this one: for other order types, you cannot merge patient records if both have active orders for the same concept
\*forgot the link

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Sorry1 I don't seem to understand your comment above

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Do you mind clarifying it please?

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: We have discussed some of the scenarios of patient merge in this [talk](https://talk.openmrs.org/t/patient-merge-functionality/9310).
I want to highlight this [part](https://talk.openmrs.org/t/patient-merge-functionality/9310/16).
We have decided the following:
 We will check only whether the patients have active orders of SAME ORDER TYPE or not.
 We will not check whether the patients have active orders of SAME CONCEPT or not, as of now.

\*I hope this is clear and forgive me for the previous incomprehensible comment.

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: I'm also commenting based on the conclusion that was made from that discussion. As per the conclusion from that discussion, I don't think the intended behavior is well reflected by the unit tests, it could be because of the way they are named, and the code changes to me don't  seem explicitly address all 3 conditions. Can you try renaming the test methods so that they are more intuitive based on the 3 conditions that need to be met.

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: There may be a gap between our understandings, since there was further discussion on the openmrs talk after the design call on 11th of Jan. Could you have a look at it once more?
The second condition of the three conditions, you are referring to, we **didn't agree** to do.

Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Well, based on the discussion it was a suggestion that was never confirmed, i have posted to the talk topic to get a green light from others. Once we get the green light, it would be nice to clean up the tests methods i.e rename them to be more intuitive to portray what they are testing more clearly and also to have more useful assertions so don't end passing simply because the test data is not well set up. E.g if a test says that it should fail if both patients have active drug orders, you need to have an assertion for each patient that ensures that they actually  have at least one drug order before proceeding with the unit test.

Comment (sidtharthanan) [inline api/src/main/java/org/openmrs/api/impl/PatientServiceImpl.java]: Great! Let us hear what people say and we will proceed.

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java (hunk 6 above)
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Can you fix this method name to say Patients**Have** and not Patient**Has**

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 9]
```diff
@@ -3304,15 +3315,43 @@ public void purgePatientIdentifierType_shouldThrowErrorWhenTryingToDeleteAPatien
...
 		expectedException.expect(APIException.class);
-		expectedException.expectMessage(Matchers.is(Context.getMessageSourceService().getMessage("Patient.cannot.merge")));
-		Patient preferredPatient = patientService.getPatient(8);
+		String message = Context.getMessageSourceService().getMessage("Patient.merge.cannotHaveSameTypeActiveOrders",
+				new Object[] { "2", "7", "Drug order" }, Context.getLocale());
+		expectedException.expectMessage(Matchers.is(message));
+		Patient preferredPatient = patientService.getPatient(2);
+		Patient notPreferredPatient = patientService.getPatient(7);
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Before merging the patients, can you add an assertion here to act as a sanity check to ensure that actually both patients have active orders of the same order type, this is helpful in case someone alters the test data and the test  ends up passing simply because the test is not well setup.

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 10]
```diff
@@ -3304,15 +3315,43 @@ public void purgePatientIdentifierType_shouldThrowErrorWhenTryingToDeleteAPatien
...
+
+	/**
+	 * @verifies not fail if one patient has active order
+	 * @see PatientService#mergePatients(org.openmrs.Patient, org.openmrs.Patient)
+	 */
+	@Test
+	public void mergePatients_shouldNotFailIfOnePatientHasActiveOrder() throws Exception {
+		Patient preferredPatient = patientService.getPatient(2);
+		Patient notPreferredPatient = patientService.getPatient(7);
+		voidOrders(Collections.singleton(notPreferredPatient));
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Same here add sanity checks which ensure that only one patient has active orders

File: api/src/test/java/org/openmrs/api/PatientServiceTest.java [hunk 11]
```diff
@@ -3304,15 +3315,43 @@ public void purgePatientIdentifierType_shouldThrowErrorWhenTryingToDeleteAPatien
...
+	/**
+	 * @verifies not fail if multiple patients has active order of different types
+	 * @see PatientService#mergePatients(org.openmrs.Patient, org.openmrs.Patient)
+	 */
+	@Test
+	public void mergePatients_shouldNotFailIfMultiplePatientsHasActiveOrderOfDifferentTypes() throws Exception {
+		Patient preferredPatient = patientService.getPatient(2);
 		Patient notPreferredPatient = patientService.getPatient(7);
+		OrderType DrugOrder = Context.getOrderService().getOrderTypeByName("Drug order");
+		voidOrdersForType(Collections.singleton(preferredPatient), DrugOrder);
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/PatientServiceTest.java]: Same here, add some sanity checks that ensure that actually both patients have active orders of different types

**Comment by sidtharthanan** [comment]
@wluyima , I've made the updates you suggested. I raised a PR [here](https://github.com/openmrs/openmrs-core/pull/1988).

**Comment by wluyima** [comment]
I've merged the new PR into 2.0.x and master, thanks @sidtharthanan
